### DISCUSSION TITLE
How to read mavens `settings.xml`?

### DISCUSSION THREAD

**Discussion by svaningelgem** [openrewrite/rewrite#4784] [Q&A] [answered]
@timtebeek : I previously filed ticket  #4589 because we couldn't use openrewrite to download internal parent pom files.
I thought it was due to the encrypted passwords (which wasn't supported yet at the time). However, this didn't resolve my issue.

So, I'm coming back with the question: how is the `settings.xml` read during a normal run when included inside maven?
As far as I could spot in the code, it's only read during tests, but not a normal run?

Could you point me to the code where it is, so I can start my investigation from there?

All I could see is that openrewrite gets only 1 repository inside the `MavenPomDownloader::downloadMetadata` method (and that is the default `repo.maven.org`, but no mirrors as we have defined).
So maybe it's also protected by a system setting (?How to configure it in the pom.xml file?), or maybe the settings file is really not taken into account?

Thanks for giving me the pointers, then I can continue my investigation into why it's still not working for us.

**Comment by timtebeek** [comment]
hi @svaningelgem ; thanks for exploring further! Know that we also have a [public Slack](https://join.slack.com/t/rewriteoss/shared_invite/zt-nj42n3ea-b~62rIHzb3Vo0E1APKCXEA) if that's an option for you. Otherwise welcome to help you here.

Did you already explore the rewrite-maven-plugin repository? That has code to configure and load the MavenSettings when we parse Maven files: https://github.com/openrewrite/rewrite-maven-plugin/blob/ab0ed4b0a21fbc689b76966bf8a9daa5468ccc45/src/main/java/org/openrewrite/maven/MavenMojoProjectParser.java#L672-L718

Could it be that your maven settings files are perhaps not in the default `~/.m2/settings.xml` location? You can also set an explicit one through our settings, as seen on the generated plugin docs: https://openrewrite.github.io/rewrite-maven-plugin/run-mojo.html

**Reply by svaningelgem** [reply]
My settings are in the default location, so that shouldn't be an issue. I can reproduce it locally with just the `rewrite` repo, no need to go to the maven-plugin one.

At least from the looks of it right now, is that the (for me at least), the `ChangeParentPom` isn't receiving any repos during the test.
Even though I changed:
```java
    public static boolean readFromDiskEnabled() {
        return true;
    }
```
So the repositories are known, just not passed down somewhere along the line.

This is the (adjusted) test that fails:
```
    @Test
    void changeParent2() {
        rewriteRun(
          spec -> spec.recipe(new ChangeParentPom(
            "com.example.adm.ws4",
            "com.example.adm.ws5",
            "adm-web-site-4-pom",
            "adm-web-site-5-pom",
            "1.0.6",
            null,
            null,
            null,
            true
          )),
          pomXml(
            """
              <project>
                <modelVersion>4.0.0</modelVersion>
              
                <parent>
                    <groupId>com.example.adm.ws4</groupId>
                    <artifactId>adm-web-site-4-pom</artifactId>
                    <version>2.1.10</version>
                </parent>
              
              
                <groupId>com.mycompany.app</groupId>
                <artifactId>my-app</artifactId>
                <version>1</version>
              </project>
              """,
            """
              <project>
                <modelVersion>4.0.0</modelVersion>
              
                <parent>
                    <groupId>com.example.adm.ws5</groupId>
                    <artifactId>adm-web-site-5-pom</artifactId>
                    <version>1.0.6</version>
                </parent>
              
                <groupId>com.mycompany.app</groupId>
                <artifactId>my-app</artifactId>
                <version>1</version>
              </project>
              """
          )
        );
    }
```

All parents referenced in the test are present in our Nexus repository. But it's simply never queried by the metadata method.
Which results in:
```
+  <!--~~(com.example.adm.ws5:adm-web-site-5-pom failed. Unable to download metadata.)~~>--><parent>
```

Any pointers would be appreciated to shorten my digging work :)

**Reply by svaningelgem** [reply]
I think it might be because `getResolutionResult().getPom().getRepositories()` isn't considering the mirrors (which are present in the context (or the `this` object)).

**Reply by timtebeek** [reply, marked as answer]
We've supported mirrors for a while, and recently added support for proxies
- https://github.com/openrewrite/rewrite/pull/7092

With that I think this has been addressed since; if not feel free to open an issue with what you need specifically.
